### PR TITLE
Log warning when Shopify indicates deprecated API call was performed

### PR DESCRIPTION
**Pull request by chrisbutcher**
Warn via the standard `ActiveResource::Base.logger` when ShopifyAPI makes a deprecated API call.

### CODE REVIEW

File: lib/active_resource/detailed_log_subscriber.rb [hunk 1]
```diff
@@ -10,8 +21,20 @@ def request(event)
...
 
-    def logger
-      ActiveResource::Base.logger
+    def warn_on_deprecated_header(event)
+      payload = event.payload
+
+      payload[:response].each do |header_name, header_value|
+        case header_name.downcase
+        when 'x-shopify-api-deprecated-reason'
+          warning_message = <<~MSG
```
Comment (eapache) [inline lib/active_resource/detailed_log_subscriber.rb]: no need for a variable here?

File: shopify_api.gemspec [hunk 2]
```diff
@@ -34,4 +34,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency("minitest", ">= 4.0")
   s.add_development_dependency("rake")
   s.add_development_dependency("timecop")
+  s.add_development_dependency("pry")
+  s.add_development_dependency("pry-byebug")
+  s.add_development_dependency("byebug")
```
Comment (jtgrenz) [inline shopify_api.gemspec]: I used to just add and remove this all the time. I wasn't sure if I should make a separate PR or sneak it into whatever I was working on.

**Comment by chrisbutcher** [comment]
Thanks. Updated!

**Comment by jtgrenz** [comment]
I'm trying to tophat this locally with a copy of Shopify_cli and I'm not getting a depreciation warning  out of the box. The nice thing about the OAuth warning is that if Im just hacking away with the cli, it spits out that warning right away. Anyway we can also log this to stdout all the time as well? Or should we  update the cli?

**Comment by chrisbutcher** [comment]
Ah interesting. I think the `warn` being called by the `OAuth.revoke` is actually Kernel#warn, which probably behaves differently than the `ActiveSupport::LogSubscriber#warn` being called in DetailedLogSubscriber in this PR.

We _could_ warn via both the Kernel and ActiveResource::Base.logger warn methods?

**Comment by jtgrenz** [comment]
I think we should. I just initialized the ActiveResource Logger in the CLI and its just too much detail. Good for a server log, not for a quick test hack.

File: lib/active_resource/detailed_log_subscriber.rb [hunk 3]
```diff
@@ -10,8 +21,18 @@ def request(event)
       info "Response:\n#{event.payload[:response].body}"
     end
 
-    def logger
-      ActiveResource::Base.logger
+    def warn_on_deprecated_header(event)
```
Comment (jtgrenz) [inline lib/active_resource/detailed_log_subscriber.rb]: maybe move this method to lib/shopify_api/connection under  `ShopifyAPI::RequestNotification.warn_on_deprecated_header` and then call `.warn_on_deprecated_header` in `.notify_about_request` 

then call again as part of `ActiveResource::DetailedLogSubscriber.log_request_response_details`?

File: test/test_helper.rb [hunk 4]
```diff
@@ -2,6 +2,7 @@
 require 'minitest/autorun'
 require 'fakeweb'
 require 'mocha/setup'
+require 'pry'
```
Comment (jamiemtdwyer) [inline test/test_helper.rb]: ✂️

Comment (chrisbutcher) [reply test/test_helper.rb]: 😢 I like pry though!

Is this really not needed? Without it, dropping `binding.pry` into code/tests fails with an undefined method error.

Comment (jtgrenz) [reply test/test_helper.rb]: agreed. Might make sense as its own PR mind you. Is there anyway we can autoload it in development so we don't need to require it and remove it? I often like to use it outside the test_helper

Comment (jamiemtdwyer) [reply test/test_helper.rb]: ah, fair enough. didn't realize your intention was to keep the dependency in, I thought this was just leftover from your debugging. I'm fine to keep it 👍

Comment (chrisbutcher) [reply test/test_helper.rb]: @jtgrenz I think if we loaded it in 'development', we'd be requiring it within client code, which we definitely don't want to do. If you're working with shopify_api in a project, that project can require 'pry' in whatever environments it wants.

File: lib/active_resource/detailed_log_subscriber.rb (hunk 3 above)
Comment (chrisbutcher) [reply lib/active_resource/detailed_log_subscriber.rb]: If I understand correctly, I think that would add coupling via direct method calls, and defeat the purpose of the pub/sub model currently being used, where `ShopifyAPI::Connection#request` ends up notifying subscribers of `request.active_resource_detailed` about the payload (req/res) it encountered.

**Comment by eapache** [comment]
Makes sense to me.

**Comment by chrisbutcher** [comment]
Looks like I need one more review from @Shopify/app-partner-dev-tools-education

**Comment by chrisbutcher** [comment]
Re-ping @Shopify/app-partner-dev-tools-education 🙏 - I just need one more review from your team to ship this, please and thanks!

**Comment by chrisbutcher** [comment]
np and thanks Tyler! <3
